source/extensions/filters/http/jwt_authn/authenticator.cc

```diff
@@ -0,0 +1,264 @@
+
+  ENVOY_LOG(debug, "Jwt authentication starts");
+  auto tokens = store_.getExtractor().extract(headers);
+  if (tokens.size() == 0) {
+    if (okToBypass()) {
+      doneWithStatus(Status::Ok);
+    } else {
+      doneWithStatus(Status::JwtMissed);
+    }
+    return;
+  }
+
+  // TODO, add supports for multiple tokens.
+  // Only process the first token for now.
+  token_.swap(tokens[0]);
+
+  Status status = jwt_.parseFromString(token_->token());
+  if (status != Status::Ok) {
+    doneWithStatus(status);
+    return;
+  }
+
+  // Check if token is extracted from the location specified by the issuer.
+  if (!token_->isIssuerSpecified(jwt_.iss_)) {
+    ENVOY_LOG(debug, "Jwt for issuer {} is not extracted from the specified locations", jwt_.iss_);
+    doneWithStatus(Status::JwtUnknownIssuer);
+    return;
+  }
+
+  // Check "exp" claim.
+  const auto unix_timestamp = std::chrono::duration_cast<std::chrono::seconds>(
+                                  std::chrono::system_clock::now().time_since_epoch())
+                                  .count();
+  if (jwt_.exp_ < unix_timestamp) {
+    doneWithStatus(Status::JwtExpired);
+    return;
+  }
+
+  // Check the issuer is configured or not.
+  jwks_data_ = store_.getJwksCache().findByIssuer(jwt_.iss_);
+  if (jwks_data_ == nullptr) {
+    doneWithStatus(Status::JwtUnknownIssuer);
+    return;
+  }
+
+  // Check if audience is allowed
+  if (!jwks_data_->areAudiencesAllowed(jwt_.audiences_)) {
+    doneWithStatus(Status::JwtAudienceNotAllowed);
+    return;
+  }
+
+  if (jwks_data_->getJwksObj() != nullptr && !jwks_data_->isExpired()) {
+    verifyKey();
+    return;
+  }
+
+  fetchRemoteJwks();
+}
+
+void AuthenticatorImpl::fetchRemoteJwks() {
+  uri_ = jwks_data_->getJwtRule().remote_jwks().http_uri().uri();
+  std::string host, path;
+  Http::Utility::extractHostPathFromUri(uri_, host, path);
+
+  Http::MessagePtr message(new Http::RequestMessageImpl());
+  message->headers().insertMethod().value().setReference(Http::Headers::get().MethodValues.Get);
+  message->headers().insertPath().value(path);
+  message->headers().insertHost().value(host);
+
+  const auto& cluster = jwks_data_->getJwtRule().remote_jwks().http_uri().cluster();
+  if (cm_.get(cluster) == nullptr) {
+    doneWithStatus(Status::JwksFetchFail);
+    return;
+  }
+
+  ENVOY_LOG(debug, "fetch pubkey from [uri = {}]: start", uri_);
+  request_ = cm_.httpAsyncClientForCluster(cluster).send(
+      std::move(message), *this, absl::optional<std::chrono::milliseconds>());
+}
+
+void AuthenticatorImpl::onSuccess(Http::MessagePtr&& response) {
+  request_ = nullptr;
+  uint64_t status_code = Http::Utility::getResponseStatus(response->headers());
+  if (status_code == 200) {
+    ENVOY_LOG(debug, "fetch pubkey [uri = {}]: success", uri_);
+    std::string body;
+    if (response->body()) {
+      auto len = response->body()->length();
+      body = std::string(static_cast<char*>(response->body()->linearize(len)), len);
```

Comment:
Nit: You can bring `std::string body` into here and make it a `const std::string`